backend/kernelCI_app/management/commands/libs/kcidb.py

```diff
@@ -0,0 +1,480 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+
+import sys
+import psycopg2
+
+
+def kcidb_execute_query(conn, query, params=None):
+    try:
+        with conn.cursor() as cur:
+            # print(cur.mogrify(query, params).decode('utf-8'))
+            cur.execute(query, params)
+            rows = cur.fetchall()
+            if not rows:
+                return []
+
+            col_names = [desc[0] for desc in cur.description]
+            result = []
+            for row in rows:
+                row_dict = dict(zip(col_names, row))
+                result.append(row_dict)
+
+            return result
+    except psycopg2.Error as e:
+        print(f"Query execution failed: {e}")
+        sys.exit()
+
+
+def kcidb_new_issues(conn, origin):
+    """Fetch issues from the last few days, including related checkouts."""
+
+    params = {"origin": origin, "interval": "4 days"}
+
+    query = """
+        WITH ranked_issues AS (
+        SELECT
+            i._timestamp,
+            i.id,
+            i.version,
+            i.comment,
+            i.misc,
+            ROW_NUMBER() OVER (PARTITION BY i.id ORDER BY i.version DESC) AS rn
+        FROM
+            public.issues i
+        WHERE origin = %(origin)s
+            AND i._timestamp >= NOW() - INTERVAL '4 days'
+        ),
+
+        highest_version AS (
+        SELECT
+            _timestamp,
+            id,
+            version,
+            comment,
+            misc
+        FROM
+            ranked_issues
+        WHERE
+            rn = 1 -- Keep only the highest version for each id
+        ),
+
+        older_issues AS (
+        SELECT h._timestamp, h.id, h.version, h.comment, h.misc
+        FROM highest_version h
+        LEFT JOIN incidents inc
+            ON h.id = inc.issue_id
+        WHERE inc._timestamp < NOW() - INTERVAL %(interval)s
+        ),
+
+        new_issues AS (
+            SELECT * FROM highest_version
+            EXCEPT
+            SELECT * FROM older_issues
+        ),
+
+        first_incidents AS (
+           SELECT
+               inc._timestamp,
+               inc.issue_id,
+               inc.issue_version,
+               inc.test_id,
+               inc.build_id,
+               c.git_repository_url,
+               c.tree_name,
+               c.git_repository_branch,
+               c.git_commit_hash,
+               c.git_commit_name,
+               ROW_NUMBER() OVER (PARTITION BY inc.issue_id ORDER BY inc._timestamp ASC) as incident_rn
+           FROM incidents inc
+           JOIN builds b ON inc.build_id = b.id
```

Comment:
Have you tried this suggestion below? I'm not sure if it would be the correct way to do it but would remove the need for an UNION, just an idea
```suggestion
           FROM incidents inc
           JOIN tests t ON inc.test_id = t.id
           JOIN builds b ON (b.id = inc.build_id OR b.id = t.build_id)
```
I think you could also use two separate joins from each table and coalesce the values of the joins into the final select